File: ipl_bbb/synthesize.R
library(tidyverse)

sanitizeData <- function(data) {
    names(data)[1:(ncol(data) - 1)] <- names(data)[2:ncol(data)]
    data[, ncol(data)] <- NULL
    sanitized <- data %>%
        na.omit() %>%
        mutate(batting_team = as.character(batting_team))  %>%
        mutate(batsman_id = as.character(batsman_id)) %>%
        mutate(batsman_name = as.character(batsman_name)) %>%
        mutate(batsman_is_rhb = as.logical(batsman_is_rhb))  %>%
        mutate(non_striker_id = as.character(non_striker_id)) %>%
        mutate(non_striker_name = as.character(non_striker_name)) %>%
        mutate(non_striker_is_rhb = as.logical(non_striker_is_rhb))  %>%
        mutate(bowling_team = as.character(bowling_team)) %>%
        mutate(bowler_id = as.character(bowler_id)) %>%
        mutate(bowler_name = as.character(bowler_name)) %>%
        mutate(bowler_is_rhb = as.logical(bowler_is_rhb)) %>%
        mutate(delivery_type = as.character(delivery_type)) %>%
        mutate(innings = as.integer(innings)) %>%
        mutate(ball = as.integer(ball)) %>%
        mutate(over = as.integer(over)) %>%
        mutate(shot_is_attacked = as.logical(shot_is_attacked)) %>%
        mutate(shot_is_played = as.logical(shot_is_played)) %>%
        mutate(shot_type = as.character(shot_type)) %>%
        mutate(bounce_above_stumps = as.character(bounce_above_stumps)) %>%
        mutate(bounce_angle = as.double(bounce_angle)) %>%
        mutate(bounce_pos_x = as.double(bounce_pos_x)) %>%
        mutate(bounce_pos_y = as.double(bounce_pos_y)) %>%
        mutate(boucne_pos_z = as.double(boucne_pos_z)) %>%
        mutate(crease_pos_x = as.double(crease_pos_x)) %>%
        mutate(crease_pos_y = as.double(crease_pos_z)) %>%
        mutate(crease_pos_z = as.double(crease_pos_z)) %>%
        mutate(deviation = as.double(deviation)) %>%
        mutate(drop_angle =  as.double(drop_angle)) %>%
        mutate(hit_stumps = as.logical(hit_stumps)) %>%
        mutate(impact_pos_x = as.double(impact_pos_x)) %>%
        mutate(impact_pos_y = as.double(impact_pos_y)) %>%
        mutate(impact_pos_z = as.double(impact_pos_z)) %>%
        mutate(initial_angle = as.double(initial_angle)) %>%
        mutate(landing_pos_x = as.double(landing_pos_x)) %>%
        mutate(landing_pos_y = as.double(landing_pos_y)) %>%
        mutate(landing_pos_z = as.double(landing_pos_z)) %>%
        mutate(pbr = as.double(pbr)) %>%
        mutate(react_time_to_crease = as.double(ifelse(react_time_to_crease == "None",0, react_time_to_crease ))) %>%
        mutate(react_time_to_intercept = as.double(ifelse(react_time_to_intercept == "None",0, react_time_to_intercept ))) %>%
        mutate(real_distance = as.double(real_distance)) %>%
        mutate(release_pos_x = as.double(release_pos_x)) %>%
        mutate(release_pos_y = as.double(release_pos_y)) %>%
        mutate(release_pos_z = as.double(release_pos_z)) %>%
        mutate(release_speed = as.double(release_speed)) %>%
        mutate(spin_rate = as.double(spin_rate)) %>%
        mutate(stump_pos_x = as.double(stump_pos_x)) %>%
        mutate(stump_pos_y = as.double(stump_pos_y)) %>%
        mutate(stump_pos_z = as.double(stump_pos_z)) %>%
        mutate(swing = as.double(swing)) %>%
        mutate(match_id = as.character(match_id)) %>%
        mutate(ball_id = as.character(ball_id)) %>%
        mutate(is_single = as.logical(is_single)) %>%
        mutate(is_double = as.logical(is_double)) %>%
        mutate(is_three = as.logical(is_three)) %>%
        mutate(is_dot = as.logical(is_dot)) %>%
        mutate(is_wide = as.logical(is_wide)) %>%
        mutate(is_no_ball = as.logical(is_no_ball)) %>%
        mutate(is_bye = as.logical(is_bye)) %>%
        mutate(is_leg_bye = as.logical(is_leg_bye)) %>%
        mutate(is_four = as.logical(is_four)) %>%
        mutate(is_six = as.logical(is_six)) %>%
        mutate(is_wicket = as.logical(is_wicket)) %>%
        mutate(wicket_type = as.character(wicket_type)) %>%
        mutate(is_bowler_wicket = as.logical(is_bowler_wicket)) %>%
        mutate(ball_type = as.character(ball_type)) %>%
        mutate(shot_type_b = as.character(shot_type_b)) %>%
        mutate(pitch_x = as.double(pitch_x)) %>%
        mutate(pitch_y = as.double(pitch_y)) %>%
        mutate(ball_line = as.character(ball_line)) %>%
        mutate(ball_length = as.character(ball_length)) %>%
        mutate(runs = as.character(runs)) %>%
        mutate(is_bouncer = as.logical(is_bouncer)) %>%
        mutate(is_free_hit = as.logical(is_free_hit)) %>%
        mutate(innings_no = as.integer(innings_no)) %>%
        mutate(ground = as.character(ground)) %>%
        mutate(date = as.character(date))

    return(sanitized)
}

data2022 <- read.csv("data2022.csv", row.names=NULL) %>%
    sanitizeData()

data2023 <- read.csv("data2023.csv", row.names=NULL) %>%
    sanitizeData()

data2024 <- read.csv("data2024.csv", row.names=NULL) %>%
    sanitizeData()

data  <- bind_rows(data2022, data2023, data2024)

write.csv(data, "data.csv", row.names=FALSE)

File: ipl_bbb/synthesize.py


import pandas as pd

def sanitize_data(data):

    df = data[data.isna().sum(axis=1) <= 42].copy()
    print("number of rows removed",len(data)-len(df))


    # Convert columns to the proper types
    df['batting_team'] = df['batting_team'].astype(str) #1 
    df['batsman_id'] = df['batsman_id'].astype(str) #2
    df['batsman_name'] = df['batsman_name'].astype(str) #3
    df['batsman_is_rhb'] = df['batsman_is_rhb'].astype(bool) #4
    df['non_striker_id'] = df['non_striker_id'].astype(str) #5
    df['non_striker_name'] = df['non_striker_name'].astype(str) #6
    df['non_striker_is_rhb'] = df['non_striker_is_rhb'].astype(bool) #7
    df['bowling_team'] = df['bowling_team'].astype(str) #8
    df['bowler_id'] = df['bowler_id'].astype(str) #9
    df['bowler_name'] = df['bowler_name'].astype(str) #10
    df['bowler_is_rhb'] = df['bowler_is_rhb'].astype(bool) #11
    df['delivery_type'] = df['delivery_type'].astype(str) #12
    # df['innings'] = df['innings'].astype(int) #13
    df['innings'] = df['innings'].astype('Int64')  # Capital I
    df['ball'] = df['ball'].astype(int) #14
    df['over'] = df['over'].astype(int) #15
    df['shot_is_attacked'] = df['shot_is_attacked'].astype(bool) #16
    df['shot_is_played'] = df['shot_is_played'].astype(bool) #17
    df['shot_type'] = df['shot_type'].astype(str) #18
    # df['bounce_above_stumps'] = df['bounce_above_stumps'].astype(str) #19
    df['bounce_angle'] = df['bounce_angle'].astype(float) #20
    df['bounce_pos_x'] = df['bounce_pos_x'].astype(float) #21
    df['bounce_pos_y'] = df['bounce_pos_y'].astype(float) #22
    df['boucne_pos_z'] = df['boucne_pos_z'].astype(float) #23
    df['crease_pos_x'] = df['crease_pos_x'].astype(float) #24
    # Note: In the original code, crease_pos_y is set from crease_pos_z.
    df['crease_pos_y'] = df['crease_pos_z'].astype(float) #25
    df['crease_pos_z'] = df['crease_pos_z'].astype(float) #26
    df['deviation'] = df['deviation'].astype(float) #27
    df['drop_angle'] = df['drop_angle'].astype(float) #28
    df['hit_stumps'] = df['hit_stumps'].astype(bool) #29
    df['impact_pos_x'] = df['impact_pos_x'].astype(float) #30
    df['impact_pos_y'] = df['impact_pos_y'].astype(float) #31
    df['impact_pos_z'] = df['impact_pos_z'].astype(float)
    df['initial_angle'] = df['initial_angle'].astype(float)
    df['landing_pos_x'] = df['landing_pos_x'].astype(float)
    df['landing_pos_y'] = df['landing_pos_y'].astype(float)
    df['landing_pos_z'] = df['landing_pos_z'].astype(float)
    df['pbr'] = df['pbr'].astype(float)
    # Replace "None" strings with 0 before converting
    df['react_time_to_crease'] = df['react_time_to_crease'].replace("None", 0).astype(float)
    df['react_time_to_intercept'] = df['react_time_to_intercept'].replace("None", 0).astype(float)
    df['real_distance'] = df['real_distance'].astype(float)
    df['release_pos_x'] = df['release_pos_x'].astype(float)
    df['release_pos_y'] = df['release_pos_y'].astype(float)
    df['release_pos_z'] = df['release_pos_z'].astype(float)
    df['release_speed'] = df['release_speed'].astype(float)
    df['spin_rate'] = df['spin_rate'].astype(float)
    df['stump_pos_x'] = df['stump_pos_x'].astype(float)
    df['stump_pos_y'] = df['stump_pos_y'].astype(float)
    df['stump_pos_z'] = df['stump_pos_z'].astype(float)
    df['swing'] = df['swing'].astype(float)
    
    df['match_id'] = df['match_id'].astype(int)
    print("the data type after conversion is ",df['match_id'].dtype)
    df['ball_id'] = df['ball_id'].astype(str)
    df['is_single'] = df['is_single'].astype(bool)
    df['is_double'] = df['is_double'].astype(bool)
    df['is_three'] = df['is_three'].astype(bool)
    df['is_dot'] = df['is_dot'].astype(bool)
    df['is_wide'] = df['is_wide'].astype(bool)
    df['is_no_ball'] = df['is_no_ball'].astype(bool)
    df['is_bye'] = df['is_bye'].astype(bool)
    df['is_leg_bye'] = df['is_leg_bye'].astype(bool)
    df['is_four'] = df['is_four'].astype(bool)
    df['is_six'] = df['is_six'].astype(bool)
    df['is_wicket'] = df['is_wicket'].astype(bool)
    df['wicket_type'] = df['wicket_type'].astype(str)
    df['is_bowler_wicket'] = df['is_bowler_wicket'].astype(bool)
    df['ball_type'] = df['ball_type'].astype(str)
    # df['shot_type_b'] = df['shot_type_b'].astype(str)
    df['pitch_x'] = df['pitch_x'].astype(float)
    df['pitch_y'] = df['pitch_y'].astype(float)
    df['ball_line'] = df['ball_line'].astype(str)
    df['ball_length'] = df['ball_length'].astype(str)
    df['runs'] = df['runs'].astype(float) 
    df['actual_runs'] = df['actual_runs'].astype(float)
    df['extras'] = df['extras'].astype(float)
    # df['ball_runs'] = df['ball_runs'].astype(float)
    df['is_bouncer'] = df['is_bouncer'].astype(bool)
    df['is_free_hit'] = df['is_free_hit'].astype(bool)
    df['innings_no'] = df['innings_no'].astype(int)
    df['ground'] = df['ground'].astype(str)
    df['date'] = df['date'].astype(str)

    return df
# Read and sanitize data
data2022 = pd.read_csv("./datasets/data2022.csv", index_col=None,low_memory=False)
#print(data2022.info())
#print(len(data2022))
data2022 = sanitize_data(data2022)
# # print(data2022 number of rows
#print(len(data2022))
data2023 = pd.read_csv("./datasets/data2023.csv", index_col=None,low_memory=False)
data2023 = sanitize_data(data2023)

data2024 = pd.read_csv("./datasets/data2024.csv", index_col=None,low_memory=False)
data2024 = sanitize_data(data2024)

data2025 = pd.read_csv("./datasets/data2025.csv", index_col=None,low_memory=False)
data2025 = sanitize_data(data2025)
# # Combine datasets
data = pd.concat([data2022, data2023, data2024,data2025], ignore_index=True)

# Write to CSV
# print(df.info())
data.to_csv("./datasets/ipl_hawkeye_data.csv", index=False)

